core/src/main/java/io/grpc/internal/DelayedClientTransport.java

```diff
@@ -497,8 +503,32 @@ private void createRealStream(ClientTransport transport) {
       setStream(realStream);
     }
 
+    // This may be called concurrently with other methods on the stream
+    private void cancelInternal(Status reason) {
+      synchronized (pendingStreamLock) {
+        if (!started) {
+          pendingCancelReason = reason;
+          return;
+        }
+      }
+      cancel(reason);
+    }
+
+    @Override
+    public final void start(ClientStreamListener listener) {
+      Status savedPendingCancelReason;
+      synchronized (pendingStreamLock) {
+        started = true;
+        savedPendingCancelReason = pendingCancelReason;
+      }
+      super.start(listener);
```

Comment:
If we're immediately cancelling, it seems pretty weak to start if it is going to do I/O. Doing unnecessary I/O in presence of cancellation is generally a recipe of cascading failure. There are two things important here though: 1) the cancellation involved here is expected to be rare and only impact a few streams since it is only due to a race between the stream creation and the delayed client transport shutdown and 2) even when it happens, no I/O is done.